# Comparison of multicultural cities

## Introduction: Description & Disscusion of the Background

New York City is one of the large cities in USA, it's also the most densely populated major city in USA. The city constitutes the premier gateway for immigration to the USA. There are more than 800 languages are spoken in New York City. It's one of the most multicultural cities in the world.
Toronto is Canada's largest city, the fourth largest in North American, and home to a diverse population. The diverse population of Toronto reflects its role as an important destination for immigrants to Canada. Over 160 languages are spoken in the city. Toronto is also one of the most multicultural cities in the world.
Los Angeles is one of the most stylish cities in the world. Almost 100 language speaking people reside at Los Angeles and hence every style, dress and cuisines of this city is different. Los Angeles is a welcoming place for immigrants.It is also one of the most multicultural cities in the world.

New York, Toronto and Los Angeles are all multicultural cities. They attract thousands of people from all over the world to live in. While the cities have some similarities and may be attractive to various people depending on lifestyles and preferences. In this project we will try to compare these three multicultural cities to see how similar they are. There are so many ways to compare cities. In this project we would like to use foursquare data to compare venues (i.e. density, distribution, categories of venues) in these cities and their clustered neighbourhoods. We will use our data science powers to determine if the cities are similar or dissimilar in multiple ways. 

People who would like to move to one of these cities may want to use the comparison results to help them make decision. In addition, people who would like to open business in these cities may use the result to help them to make decisions. 

## Data Description

Based on definition of our problem, factors that will influence our decision are:
<br>
•	How many categories of venues in the neighborhood/city?
<br>
•	Density of venues in the neighborhood/city?
<br>
•	What categories of venues are most popular in each city?
<br>
We can use these factors to determine if cities are similar to each other.

To consider the problem we can list the data as below:
<br>
I use https://geo.nyu.edu/catalog/nyu_2451_34572 data to get neighborhoods as well as the the latitude and logitude coordinates of each neighborhood of New York City.Example:  We can find Toronto's neibourhoods and the latitude and logitude coordinates of each neighborhood from wikipedia https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M. A csv file contians latitude and logitude information of Toronto. Los Angeles neibourhoods information are available on SOCR Data LA Neighborhoods: Datahttp://wiki.stat.ucla.edu/socr/index.php/SOCR_Data_LA_Neighborhoods_Data#SOCR_Data_-_Los_Angeles_City_Neighborhoods_Data.

I used Forsquare API to get the most common venues of given neibourhoods of three cities.

In [3]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


### Download and Cleaning New York City Data

Let's first find the latitude & longitude of new york city neibourhoods.

In [4]:
!wget -q -O 'newyork_data.json' https://cocl.us/new_york_dataset
print('Data downloaded!')

Data downloaded!


In [5]:
with open('newyork_data.json') as json_data:
    newyork_data = json.load(json_data)

In [6]:
ny_neighborhoods_data = newyork_data['features']
# define the dataframe columns
column_names = ['Borough', 'Neighborhood', 'Latitude', 'Longitude'] 

# instantiate the dataframe
ny_neighborhoods = pd.DataFrame(columns=column_names)

In [7]:
for data in ny_neighborhoods_data:
    borough = neighborhood_name = data['properties']['borough']
    neighborhood_name = data['properties']['name']
        
    neighborhood_latlon = data['geometry']['coordinates']
    neighborhood_lat = neighborhood_latlon[1]
    neighborhood_lon = neighborhood_latlon[0]
    
    ny_neighborhoods = ny_neighborhoods.append({'Borough': borough,
                                                'Neighborhood': neighborhood_name,
                                                'Latitude': neighborhood_lat,
                                                'Longitude': neighborhood_lon}, ignore_index=True)

In [35]:
ny_neighborhoods = ny_neighborhoods[['Neighborhood','Latitude','Longitude']]

In [9]:
ny_neighborhoods.head(5)

Neighborhood   Latitude  Longitude
0    Wakefield  40.894705 -73.847201
1   Co-op City  40.874294 -73.829939
2  Eastchester  40.887556 -73.827806
3    Fieldston  40.895437 -73.905643
4    Riverdale  40.890834 -73.912585

### Download and Cleaning Toronto Data

Then find the latitude & longitude of Toronto neibourhoods.

In [10]:
toronto_df = pd.read_html("https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M")[0]
# remove rows if Borough is Not assigned
toronto_df = toronto_df[toronto_df.Borough != 'Not assigned']
# rows will be combined into one row with the neighborhoods separated with a comma if Postcode is the same
df1=toronto_df.groupby('Postcode').agg({'Borough':'first',
                               'Neighbourhood': ', '.join}).reset_index()
# If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough
df1.loc[df1['Neighbourhood'] =='Not assigned', 'Neighbourhood'] = df1['Borough']

In [11]:
# read in csv file
df_part = pd.read_csv("http://cocl.us/Geospatial_data/Geospatial_Coordinates.csv")
df_part=df_part.rename(columns={'Postal Code':'Postcode','Latitude':'Latitude','Longitude':'Longitude'})

In [12]:
# append latitude and longtitude to postcode
df_toronto = pd.merge(df1,df_part,on='Postcode')

In [13]:
df_toronto = df_toronto[['Neighbourhood','Latitude','Longitude']]

In [14]:
df_toronto.head(5)

Neighbourhood   Latitude  Longitude
0                          Rouge, Malvern  43.806686 -79.194353
1  Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497
2       Guildwood, Morningside, West Hill  43.763573 -79.188711
3                                  Woburn  43.770992 -79.216917
4                               Cedarbrae  43.773136 -79.239476

### Download and Cleaning Los Angeles Data

Find the latitude & longitude of Los Angeles neibourhoods.

In [15]:
Los_df = pd.read_html("http://wiki.stat.ucla.edu/socr/index.php/SOCR_Data_LA_Neighborhoods_Data#SOCR_Data_-_Los_Angeles_City_Neighborhoods_Data")[2]

In [16]:
Los_df = Los_df[['LA_Nbhd','Longitude','Latitude']]

In [17]:
Los_df.columns = ['Neighbourhood','Longitude','Latitude']

In [18]:
Los_df.head(5)

Neighbourhood   Longitude   Latitude
0         Adams_Normandie -118.300270  34.030970
1                  Arleta -118.430015  34.240603
2       Arlington_Heights -118.320109  34.043611
3         Atwater_Village -118.265808  34.124908
4  Baldwin_Hills/Crenshaw -118.366700  34.019090

## Methodology

In this project we will direct our efforts on comparing similiartity in pairs. 

In first step we have collected the required data: latitude, longitude, name and type (category) of every venues within 500 meters and limit of 100 venues from center of each neibourhood in each city. 

Second step in our analysis will be calculation and exploration of venue density across different cities - we will find out how many types of venues in each city and density of each type of venues in each city. We can cluster and plot the cluster 
of neibourhoods in each city. 

In third and final step we will use cosine similarity to calculate how similar the cities in pairs. Mathematically, cosine similarity measures the cosine of the angle between two vectors projected in a multi-dimensional space. The types and number of venues under each type in each city can be consider as a vector. We put three cities in to three pairs to compare similarity between two cities each time, then find which two cities are most similar to each other from these three cities: New York, Toronto and Los Angeles. 

## Analysis

### Data from Foursquare API

Now that we have our location candidates, let's use Foursquare API to get info on venues in each neighborhood in each city.
We're interested in venues in 500 meters radius of each neibourhood and limit of 100 venues in each neighborhood in all the cities.

In [19]:
# Define Foursquare Credentials and Version
CLIENT_ID = 'JUVSE5RRWDONTWQJQJ4C0AGCZQHTM5UNXNWV3W5PLLBMCVH1' # your Foursquare ID
CLIENT_SECRET = 'GT1QAOAKGKIGQJJSBMFOQ0ZDCWTPOJDAQXNMVACNXZCREH0Z' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

### Get venues in New York City

In [20]:
# let's get the top 100 venues that are in each neibourhood of New York City within a radius of 500 meters
LIMIT = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [22]:
ny_venues.head(5)

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0     Wakefield               40.894705               -73.847201   
1     Wakefield               40.894705               -73.847201   
2     Wakefield               40.894705               -73.847201   
3     Wakefield               40.894705               -73.847201   
4     Wakefield               40.894705               -73.847201   

                                     Venue  Venue Latitude  Venue Longitude  \
0                         Lollipops Gelato       40.894123       -73.845892   
1                                 Rite Aid       40.896649       -73.844846   
2                         Carvel Ice Cream       40.890487       -73.848568   
3  Cooler Runnings Jamaican Restaurant Inc       40.898276       -73.850381   
4                                    Shell       40.894187       -73.845862   

         Venue Category  
0          Dessert Shop  
1              Pharmacy  
2        Ice Cream Shop  
3  Caribbean Restaurant  
4           Gas Station

In [23]:
# analyze each Neighborhood in New York City
# one hot encoding
ny_onehot = pd.get_dummies(ny_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
ny_onehot['Neighbourhood'] = ny_venues['Neighbourhood'] 

# move neighborhood column to the first column
fixed_columns = [ny_onehot.columns[-1]] + list(ny_onehot.columns[:-1])
ny_onehot = ny_onehot[fixed_columns]

ny_onehot.head(5)

Neighbourhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0     Wakefield                  0               0                  0   
1     Wakefield                  0               0                  0   
2     Wakefield                  0               0                  0   
3     Wakefield                  0               0                  0   
4     Wakefield                  0               0                  0   

   African Restaurant  Airport Terminal  American Restaurant  Animal Shelter  \
0                   0                 0                    0               0   
1                   0                 0                    0               0   
2                   0                 0                    0               0   
3                   0                 0                    0               0   
4                   0                 0                    0               0   

   Antique Shop  Arcade  Arepa Restaurant  Argentinian Restaurant  \
0             0       0                 0                       0   
1             0       0                 0                       0   
2             0       0                 0                       0   
3             0       0                 0                       0   
4             0       0                 0                       0   

   Art Gallery  Art Museum  Arts & Crafts Store  Arts & Entertainment  \
0            0           0                    0                     0   
1            0           0                    0                     0   
2            0           0                    0                     0   
3            0           0                    0                     0   
4            0           0                    0                     0   

   Asian Restaurant  Athletics & Sports  Auditorium  Australian Restaurant  \
0                 0                   0           0                      0   
1                 0                   0           0                      0   
2                 0                   0           0                      0   
3                 0                   0           0                      0   
4                 0                   0           0                      0   

   Austrian Restaurant  Auto Garage  Auto Workshop  Automotive Shop  \
0                    0            0              0                0   
1                    0            0              0                0   
2                    0            0              0                0   
3                    0            0              0                0   
4                    0            0              0                0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    0               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Bath House  Beach  Beach Bar  \
0                 0                 0           0      0          0   
1                 0                 0           0      0          0   
2                 0                 0           0      0          0   
3                 0                 0           0      0          0   
4                 0                 0           0      0          0   

   Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0                0         0            0           0              0   
1                0         0            0           0              0   
2                0         0            0           0              0   
3                0         0            0           0              0   
4                0         0            0           0              0   

   Bike Rental / Bike

In [24]:
ny_onehot.shape

(10385, 427)

There is 426 categories of venues in New York City

In [25]:
# count number of categories of venues in each neiborhood
ny_onehot_group=ny_onehot.groupby(['Neighbourhood']).sum()
ny_onehot_group.head()

Accessories Store  Adult Boutique  Afghan Restaurant  \
Neighbourhood                                                         
Allerton                       0               0                  0   
Annadale                       0               0                  0   
Arden Heights                  0               0                  0   
Arlington                      0               0                  0   
Arrochar                       0               0                  0   

               African Restaurant  Airport Terminal  American Restaurant  \
Neighbourhood                                                              
Allerton                        0                 0                    1   
Annadale                        0                 0                    1   
Arden Heights                   0                 0                    0   
Arlington                       0                 0                    1   
Arrochar                        0                 0                    0   

               Animal Shelter  Antique Shop  Arcade  Arepa Restaurant  \
Neighbourhood                                                           
Allerton                    0             0       0                 0   
Annadale                    0             0       0                 0   
Arden Heights               0             0       0                 0   
Arlington                   0             0       0                 0   
Arrochar                    0             0       0                 0   

               Argentinian Restaurant  Art Gallery  Art Museum  \
Neighbourhood                                                    
Allerton                            0            0           0   
Annadale                            0            0           0   
Arden Heights                       0            0           0   
Arlington                           0            0           0   
Arrochar                            0            0           0   

               Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
Neighbourhood                                                                
Allerton                         0                     0                 0   
Annadale                         0                     0                 0   
Arden Heights                    0                     0                 0   
Arlington                        0                     0                 0   
Arrochar                         0                     0                 0   

               Athletics & Sports  Auditorium  Australian Restaurant  \
Neighbourhood                                                          
Allerton                        0           0                      0   
Annadale                        0           0                      0   
Arden Heights                   0           0                      0   
Arlington                       0           0                      0   
Arrochar                        1           0                      0   

               Austrian Restaurant  Auto Garage  Auto Workshop  \
Neighbourhood                                                    
Allerton                         0            0              0   
Annadale                         0            0              0   
Arden Heights                    0            0              0   
Arlington                        0            0              0   
Arrochar                         0            0              0   

               Automotive Shop  BBQ Joint  Baby Store  Bagel Shop  Bakery  \
Neighbourhood                                                               
Allerton                     0          0           0           0       1   
Annadale                     0          0           0           1       0   
Arden Heights                0          0           0           0       0   
Arlington                    0          0           0           0       0   
Arrochar                     0  

In [26]:
# Calculate the total per each category of venue in New York City
ny_city=pd.DataFrame(data=ny_onehot_group.sum(axis=0))

In [27]:
ny_city=ny_city.transpose()

In [28]:
ny_city.rename(index = {0:"New York"},inplace=True)

We would like to calculate number of venues under each type in New York City.

In [29]:
# create a dataframe which contains the sum of venues by category in New York City
ny_city

Accessories Store  Adult Boutique  Afghan Restaurant  \
New York                 12               3                  3   

          African Restaurant  Airport Terminal  American Restaurant  \
New York                   6                 1                  169   

          Animal Shelter  Antique Shop  Arcade  Arepa Restaurant  \
New York               1            13       3                10   

          Argentinian Restaurant  Art Gallery  Art Museum  \
New York                       9           55           7   

          Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
New York                   15                     2                51   

          Athletics & Sports  Auditorium  Australian Restaurant  \
New York                  16           1                      4   

          Austrian Restaurant  Auto Garage  Auto Workshop  Automotive Shop  \
New York                    2            1              3                3   

          BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
New York         27           1         106     231   154  210   

          Baseball Field  Baseball Stadium  Basketball Court  Bath House  \
New York              20                 3                13           1   

          Beach  Beach Bar  Bed & Breakfast  Beer Bar  Beer Garden  \
New York     45          1                2        13           14   

          Beer Store  Big Box Store  Bike Rental / Bike Share  Bike Shop  \
New York           6              5                         6          4   

          Bike Trail  Bistro  Board Shop  Boat or Ferry  Bookstore  Boutique  \
New York           3       8          10             10         52        38   

          Bowling Alley  Boxing Gym  Brazilian Restaurant  Breakfast Spot  \
New York             10          10                     5              38   

          Brewery  Bridal Shop  Bridge  Bubble Tea Shop  Buffet  Building  \
New York       14            3       3               30       4         8   

          Burger Joint  Burrito Place  Bus Line  Bus Station  Bus Stop  \
New York            79              9        10           72        77   

          Business Service  Butcher  Cafeteria  Café  \
New York                 8        7          2   164   

          Cajun / Creole Restaurant  Cambodian Restaurant  Camera Store  \
New York                         10                     1             2   

          Campground  Candy Store  Cantonese Restaurant  Caribbean Restaurant  \
New York           1           13                     4                    85   

          Caucasian Restaurant  Cha Chaan Teng  Check Cashing Service  \
New York                     3               1                      8   

          Cheese Shop  Child Care Service  Chinese Restaurant  Chiropractor  \
New York            9                   2                 222             1   

          Chocolate Shop  Church  Climbing Gym  Clothing Store  Club House  \
New York               5       2             4              79           2   

          Cocktail Bar  Coffee Shop  College Academic Building  \
New York           101          275                          2   

          College Basketball Court  College Bookstore  College Cafeteria  \
New York                         1                  1                  2   

          College Theater  Colombian Restaurant  Comedy Club  \
New York                1                     3            7   

          Comfort Food Restaurant  Comic Shop  Community Center  Concert Hall  \
New York                        3           2                 2            14   

          Construction & Landscaping  Convenience Store  Cooking School  \
New York                          11                 51               1   

          Cosmetics Shop  Coworking Space  Creperie  Cuban Restaurant  \
New York              93                3         9                17   

          Cupcake Shop  Cycle Studio  Czech Restaurant  Dance Studio 

In [30]:
#  group rows by neighborhood and by taking the mean of the frequency of occurrence of each category in New York City
ny_grouped_sum = ny_onehot.groupby('Neighbourhood').mean().reset_index()
ny_grouped_sum.head(5)

Neighbourhood  Accessories Store  Adult Boutique  Afghan Restaurant  \
0       Allerton                0.0             0.0                0.0   
1       Annadale                0.0             0.0                0.0   
2  Arden Heights                0.0             0.0                0.0   
3      Arlington                0.0             0.0                0.0   
4       Arrochar                0.0             0.0                0.0   

   African Restaurant  Airport Terminal  American Restaurant  Animal Shelter  \
0                 0.0               0.0             0.032258             0.0   
1                 0.0               0.0             0.076923             0.0   
2                 0.0               0.0             0.000000             0.0   
3                 0.0               0.0             0.111111             0.0   
4                 0.0               0.0             0.000000             0.0   

   Antique Shop  Arcade  Arepa Restaurant  Argentinian Restaurant  \
0           0.0     0.0               0.0                     0.0   
1           0.0     0.0               0.0                     0.0   
2           0.0     0.0               0.0                     0.0   
3           0.0     0.0               0.0                     0.0   
4           0.0     0.0               0.0                     0.0   

   Art Gallery  Art Museum  Arts & Crafts Store  Arts & Entertainment  \
0          0.0         0.0                  0.0                   0.0   
1          0.0         0.0                  0.0                   0.0   
2          0.0         0.0                  0.0                   0.0   
3          0.0         0.0                  0.0                   0.0   
4          0.0         0.0                  0.0                   0.0   

   Asian Restaurant  Athletics & Sports  Auditorium  Australian Restaurant  \
0               0.0            0.000000         0.0                    0.0   
1               0.0            0.000000         0.0                    0.0   
2               0.0            0.000000         0.0                    0.0   
3               0.0            0.000000         0.0                    0.0   
4               0.0            0.058824         0.0                    0.0   

   Austrian Restaurant  Auto Garage  Auto Workshop  Automotive Shop  \
0                  0.0          0.0            0.0              0.0   
1                  0.0          0.0            0.0              0.0   
2                  0.0          0.0            0.0              0.0   
3                  0.0          0.0            0.0              0.0   
4                  0.0          0.0            0.0              0.0   

   BBQ Joint  Baby Store  Bagel Shop    Bakery  Bank  Bar  Baseball Field  \
0        0.0         0.0    0.000000  0.032258   0.0  0.0             0.0   
1        0.0         0.0    0.076923  0.000000   0.0  0.0             0.0   
2        0.0         0.0    0.000000  0.000000   0.0  0.0             0.0   
3        0.0         0.0    0.000000  0.000000   0.0  0.0             0.0   
4        0.0         0.0    0.058824  0.000000   0.0  0.0             0.0   

   Baseball Stadium  Basketball Court  Bath House  Beach  Beach Bar  \
0               0.0               0.0         0.0    0.0        0.0   
1               0.0               0.0         0.0    0.0        0.0   
2               0.0               0.0         0.0    0.0        0.0   
3               0.0               0.0         0.0    0.0        0.0   
4               0.0               0.0         0.0    0.0        0.0   

   Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  Big Box Store  \
0              0.0       0.0          0.0         0.0            0.0   
1              0.0       0.0          0.0         0.0            0.0   
2              0.0       0.0          0.0         0.0            0.0   
3              0.0       0.0          0.0         0.0            0.0   
4              0.0       0.0          0.0         0.0            0.0   

   B

find the top 10 venues for each neighborhood in New York City

In [59]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [60]:
# create the new dataframe and display the top 10 venues for each neighborhood
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
ny_neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
ny_neighborhoods_venues_sorted['Neighbourhood'] = ny_grouped_sum['Neighbourhood']

for ind in np.arange(ny_grouped_sum.shape[0]):
    ny_neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(ny_grouped_sum.iloc[ind, :], num_top_venues)

ny_neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0       Allerton         Deli / Bodega           Pizza Place   
1       Annadale              Pharmacy                   Pub   
2  Arden Heights           Coffee Shop              Bus Stop   
3      Arlington              Bus Stop         Grocery Store   
4       Arrochar    Italian Restaurant         Deli / Bodega   

  3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
0                   Spa           Supermarket                   Pharmacy   
1         Train Station           Pizza Place                 Bagel Shop   
2           Pizza Place              Pharmacy                Yoga Studio   
3         Deli / Bodega          Intersection        American Restaurant   
4              Bus Stop                 Hotel  Middle Eastern Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Playground            Bike Trail     Martial Arts Dojo   
1                 Diner                  Food            Sports Bar   
2                 Field         Event Service           Event Space   
3                  Food         Boat or Ferry           Coffee Shop   
4            Bagel Shop        Sandwich Place           Pizza Place   

  9th Most Common Venue    10th Most Common Venue  
0  Fast Food Restaurant               Gas Station  
1             Pet Store       American Restaurant  
2               Exhibit                Eye Doctor  
3           Yoga Studio      Fast Food Restaurant  
4    Athletics & Sports  Mediterranean Restaurant

Run k-means to cluster the neighborhood into 5 clusters.

In [61]:
# cluster Neighbourhood
# set number of clusters
kclusters = 5

ny_grouped_clustering = ny_grouped_sum.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(ny_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([2, 2, 2, 0, 0, 0, 0, 2, 0, 2], dtype=int32)

In [41]:
ny_neighborhoods1=ny_neighborhoods.rename(columns={"Neighborhood": "Neighbourhood"})

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [62]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood
# add clustering labels
ny_neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
ny_merged = ny_neighborhoods1
ny_merged = ny_merged.join(ny_neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

ny_merged.head()

Neighbourhood   Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0     Wakefield  40.894705 -73.847201             2.0        Ice Cream Shop   
1    Co-op City  40.874294 -73.829939             2.0           Bus Station   
2   Eastchester  40.887556 -73.827806             2.0  Caribbean Restaurant   
3     Fieldston  40.895437 -73.905643             2.0                 River   
4     Riverdale  40.890834 -73.912585             2.0                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Dessert Shop        Sandwich Place           Pizza Place   
1     Accessories Store        Discount Store  Fast Food Restaurant   
2         Deli / Bodega                 Diner         Metro Station   
3           Bus Station            Playground                 Plaza   
4                  Bank            Playground                   Gym   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Laundromat  Caribbean Restaurant           Gas Station   
1             Gift Shop                  Park         Grocery Store   
2         Bowling Alley    Seafood Restaurant  Fast Food Restaurant   
3           Yoga Studio                 Field         Event Service   
4             Locksmith          Home Service                 Plaza   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Donut Shop              Pharmacy             Food Truck  
1        Baseball Field        Mattress Store            Pizza Place  
2           Pizza Place              Platform             Donut Shop  
3           Event Space               Exhibit             Eye Doctor  
4            Food Truck           Bus Station      Food & Drink Shop

In [65]:
ny_merged = ny_merged.fillna(0)

In [66]:
ny_merged['Cluster Labels'] = ny_merged['Cluster Labels'].apply(np.int32)

In [68]:
# visualize the resulting clusters
# create map
latitude = 40.730610
longitude = -73.935242
ny_map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(ny_merged['Latitude'], ny_merged['Longitude'], ny_merged['Neighbourhood'], ny_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(ny_map_clusters)
       
ny_map_clusters

### Get venues in Toronto

In [69]:
# Toronto venues
tor_venues = getNearbyVenues(names=df_toronto['Neighbourhood'],
                                   latitudes=df_toronto['Latitude'],
                                   longitudes=df_toronto['Longitude']
                                  )
print(tor_venues.shape)

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

In [70]:
tor_venues.head(5)

Neighbourhood  Neighbourhood Latitude  \
0                          Rouge, Malvern               43.806686   
1  Highland Creek, Rouge Hill, Port Union               43.784535   
2  Highland Creek, Rouge Hill, Port Union               43.784535   
3       Guildwood, Morningside, West Hill               43.763573   
4       Guildwood, Morningside, West Hill               43.763573   

   Neighbourhood Longitude                            Venue  Venue Latitude  \
0               -79.194353                          Wendy's       43.807448   
1               -79.160497            Royal Canadian Legion       43.782533   
2               -79.160497   Scarborough Historical Society       43.788755   
3               -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
4               -79.188711                G & G Electronics       43.765309   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.162438        History Museum  
3       -79.189914           Pizza Place  
4       -79.191537     Electronics Store

In [71]:
# analyze each Neighbourhood in Toronto
# one hot encoding
tor_onehot = pd.get_dummies(tor_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighbourhood column back to dataframe
tor_onehot['Neighbourhood'] = tor_venues['Neighbourhood'] 

# move neighbourhood column to the first column
fixed_columns = [tor_onehot.columns[-1]] + list(tor_onehot.columns[:-1])
tor_onehot = tor_onehot[fixed_columns]

tor_onehot.head(5)

Neighbourhood  Accessories Store  \
0                          Rouge, Malvern                  0   
1  Highland Creek, Rouge Hill, Port Union                  0   
2  Highland Creek, Rouge Hill, Port Union                  0   
3       Guildwood, Morningside, West Hill                  0   
4       Guildwood, Morningside, West Hill                  0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0                  0        0                   0               0   
1                  0        0                   0               0   
2                  0        0                   0               0   
3                  0        0                   0               0   
4                  0        0                   0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Field  Baseball Stadium  Basketball Stadium  \
0       0     0    0               0                 0                   0   
1       0     0    1               0                 0                   0   
2       0     0    0               0                 0                   0   
3       0     0    0               0                 0                   0   
4       0     0    0               0                 0                   0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0      0                0         0           0                   0   
1      0                0         0           0                   0   
2      0                0         0           0                   0   
3      0                0         0           0                   0   
4      0                0         0           0                   0   

   Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0          0       0              0          0         0   
1          0       0              0          0         0   
2          0       0              0          0         0   
3          0       0              0          0         0   
4          0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                     0               0        0            0   
1                     0               0        0            0   
2                     0               0        0            0   
3                     0               0        0            0   
4                     0               0        0            0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0                0         0             0              0         0   
1                0         0             0              0    

In [72]:
tor_onehot.shape

(2281, 276)

There is 275 categories of venues in Toronto

In [110]:
# count number of categories of venues in each neiborhood
tor_onehot_group=tor_onehot.groupby(['Neighbourhood']).sum()
tor_onehot_group.head()

Accessories Store  \
Neighbourhood                                                           
Adelaide, King, Richmond                                            0   
Agincourt                                                           0   
Agincourt North, L'Amoreaux East, Milliken, Ste...                  0   
Albion Gardens, Beaumond Heights, Humbergate, J...                  0   
Alderwood, Long Branch                                              0   

                                                    Afghan Restaurant  \
Neighbourhood                                                           
Adelaide, King, Richmond                                            0   
Agincourt                                                           0   
Agincourt North, L'Amoreaux East, Milliken, Ste...                  0   
Albion Gardens, Beaumond Heights, Humbergate, J...                  0   
Alderwood, Long Branch                                              0   

                                                    Airport  \
Neighbourhood                                                 
Adelaide, King, Richmond                                  0   
Agincourt                                                 0   
Agincourt North, L'Amoreaux East, Milliken, Ste...        0   
Albion Gardens, Beaumond Heights, Humbergate, J...        0   
Alderwood, Long Branch                                    0   

                                                    Airport Food Court  \
Neighbourhood                                                            
Adelaide, King, Richmond                                             0   
Agincourt                                                            0   
Agincourt North, L'Amoreaux East, Milliken, Ste...                   0   
Albion Gardens, Beaumond Heights, Humbergate, J...                   0   
Alderwood, Long Branch                                               0   

                                                    Airport Lounge  \
Neighbourhood                                                        
Adelaide, King, Richmond                                         0   
Agincourt                                                        0   
Agincourt North, L'Amoreaux East, Milliken, Ste...               0   
Albion Gardens, Beaumond Heights, Humbergate, J...               0   
Alderwood, Long Branch                                           0   

                                                    Airport Service  \
Neighbourhood                                                         
Adelaide, King, Richmond                                          0   
Agincourt                                                         0   
Agincourt North, L'Amoreaux East, Milliken, Ste...                0   
Albion Gardens, Beaumond Heights, Humbergate, J...                0   
Alderwood, Long Branch                                            0   

                                                    Airport Terminal  \
Neighbourhood                                                          
Adelaide, King, Richmond                                           0   
Agincourt                                                          0   
Agincourt North, L'Amoreaux East, Milliken, Ste...                 0   
Albion Gardens, Beaumond Heights, Humbergate, J...                 0   
Alderwood, Long Branch                                             0   

                                                    American Restaurant  \
Neighbourhood                                                             
Adelaide, King, Richmond                                              3   
Agincourt                                                             0   
Agincourt North, L'Amoreaux East, Milliken, Ste...                    0   
Albion Gardens, Beaumond Heights, Humbergate, J...                    0   
Alderwood, Long Branch                                                0   

                                               

In [74]:
# Calculate the total per each category of venue in Toronto
tor_city=pd.DataFrame(data=tor_onehot_group.sum(axis=0))

In [75]:
tor_city=tor_city.transpose()

In [76]:
tor_city.rename(index = {0:"Toronto"},inplace=True)

We would like to calculate number of venues under each type in Toronto.

In [77]:
# create a dataframe which contains the sum of venues by category in Toronto
tor_city

Accessories Store  Afghan Restaurant  Airport  Airport Food Court  \
Toronto                  1                  1        2                   1   

         Airport Lounge  Airport Service  Airport Terminal  \
Toronto               2                2                 2   

         American Restaurant  Antique Shop  Aquarium  Art Gallery  Art Museum  \
Toronto                   29             2         5           13           2   

         Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
Toronto                    7                17                   6   

         Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
Toronto              1          5           1           7      49    18   42   

         Baseball Field  Baseball Stadium  Basketball Stadium  Beach  \
Toronto               5                 2                   4      1   

         Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  \
Toronto                1        18           9                   3          2   

         Bistro  Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  \
Toronto       4              1         18         5                     3   

         Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
Toronto              25       14            1               10         2   

         Burger Joint  Burrito Place  Bus Line  Bus Station  Business Service  \
Toronto            28             13         6            3                 1   

         Butcher  Café  Cajun / Creole Restaurant  Camera Store  Candy Store  \
Toronto        3   105                          1             1            1   

         Caribbean Restaurant  Cheese Shop  Chinese Restaurant  \
Toronto                    10            7                  23   

         Chocolate Shop  Church  Climbing Gym  Clothing Store  Cocktail Bar  \
Toronto               1       2             1              35            14   

         Coffee Shop  College Arts Building  College Auditorium  College Gym  \
Toronto          193                      1                   1            1   

         College Rec Center  College Stadium  Colombian Restaurant  \
Toronto                   1                1                     2   

         Comfort Food Restaurant  Comic Shop  Concert Hall  \
Toronto                        8           3            10   

         Construction & Landscaping  Convenience Store  Cosmetics Shop  \
Toronto                           2                 10              17   

         Coworking Space  Creperie  Cuban Restaurant  Cupcake Shop  \
Toronto                1         7                 2             2   

         Curling Ice  Dance Studio  Deli / Bodega  Department Store  \
Toronto            1             4             16                 8   

         Dessert Shop  Dim Sum Restaurant  Diner  Discount Store  Dog Run  \
Toronto            16                   3     20              12        5   

         Doner Restaurant  Donut Shop  Drugstore  Dumpling Restaurant  \
Toronto                 1           3          1                    4   

         Eastern European Restaurant  Electronics Store  Empanada Restaurant  \
Toronto                            4                  9                    1   

         Ethiopian Restaurant  Event Space  Falafel Restaurant  \
Toronto                     2            2                   3   

         Farmers Market  Fast Food Restaurant  Festival  Field  \
Toronto              11                    35         1      1   

         Filipino Restaurant  Fish & Chips Shop  Fish Market  Flea Market  \
Toronto                    1                  4            5            1   

         Flower Shop  Food  Food & Drink Shop  Food Court  Food Truck  \
Toronto            2     1                  4           9           4   

         Fountain  French Restaurant  Fried Chicken Joint  Frozen Yogurt Shop  \
Toronto         4                 11                   16 

In [78]:
#  group rows by neighborhood and by taking the mean of the frequency of occurrence of each category in Toronto
tor_grouped_sum = tor_onehot.groupby('Neighbourhood').mean().reset_index()
tor_grouped_sum.head(5)

Neighbourhood  Accessories Store  \
0                           Adelaide, King, Richmond                0.0   
1                                          Agincourt                0.0   
2  Agincourt North, L'Amoreaux East, Milliken, St...                0.0   
3  Albion Gardens, Beaumond Heights, Humbergate, ...                0.0   
4                             Alderwood, Long Branch                0.0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Lounge  \
0                0.0      0.0                 0.0             0.0   
1                0.0      0.0                 0.0             0.0   
2                0.0      0.0                 0.0             0.0   
3                0.0      0.0                 0.0             0.0   
4                0.0      0.0                 0.0             0.0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0              0.0               0.0                 0.03           0.0   
1              0.0               0.0                 0.00           0.0   
2              0.0               0.0                 0.00           0.0   
3              0.0               0.0                 0.00           0.0   
4              0.0               0.0                 0.00           0.0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0       0.0         0.01        0.01                  0.0          0.020000   
1       0.0         0.00        0.00                  0.0          0.000000   
2       0.0         0.00        0.00                  0.0          0.333333   
3       0.0         0.00        0.00                  0.0          0.000000   
4       0.0         0.00        0.00                  0.0          0.000000   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0            0.000000            0.0        0.0         0.0         0.0   
1            0.000000            0.0        0.0         0.0         0.0   
2            0.000000            0.0        0.0         0.0         0.0   
3            0.000000            0.0        0.0         0.0         0.0   
4            0.090909            0.0        0.0         0.0         0.0   

   Bakery  Bank   Bar  Baseball Field  Baseball Stadium  Basketball Stadium  \
0    0.02   0.0  0.04             0.0               0.0                 0.0   
1    0.00   0.0  0.00             0.0               0.0                 0.0   
2    0.00   0.0  0.00             0.0               0.0                 0.0   
3    0.00   0.0  0.00             0.0               0.0                 0.0   
4    0.00   0.0  0.00             0.0               0.0                 0.0   

   Beach  Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  \
0    0.0              0.0       0.0    0.000000                 0.0   
1    0.0              0.0       0.0    0.000000                 0.0   
2    0.0              0.0       0.0    0.000000                 0.0   
3    0.0              0.0       0.0    0.090909                 0.0   
4    0.0              0.0       0.0    0.000000                 0.0   

   Bike Shop  Bistro  Boat or Ferry  Bookstore  Boutique  \
0        0.0     0.0            0.0       0.01       0.0   
1        0.0     0.0            0.0       0.00       0.0   
2        0.0     0.0            0.0       0.00       0.0   
3        0.0     0.0            0.0       0.00       0.0   
4        0.0     0.0            0.0       0.00       0.0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bridal Shop  \
0                  0.01            0.02      0.0          0.0   
1                  0.00            0.25      0.0          0.0   
2                  0.00            0.00      0.0          0.0   
3                  0.00            0.00      0.0          0.0   
4                  0.00            0.00      0.0          0.0   

   Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0              0.0      0.01          0.03           0.01       0.0   


In [79]:
# create a new dataframe
tor_neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
tor_neighborhoods_venues_sorted['Neighbourhood'] = tor_grouped_sum['Neighbourhood']

for ind in np.arange(tor_grouped_sum.shape[0]):
    tor_neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(tor_grouped_sum.iloc[ind, :], num_top_venues)

tor_neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt        Breakfast Spot   
2  Agincourt North, L'Amoreaux East, Milliken, St...            Playground   
3  Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café                   Bar            Steakhouse   
1                Lounge        Clothing Store        Sandwich Place   
2                  Park      Asian Restaurant           Yoga Studio   
3              Pharmacy        Sandwich Place   Japanese Restaurant   
4          Skating Rink        Sandwich Place          Dance Studio   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Cosmetics Shop          Burger Joint       Thai Restaurant   
1              Festival  Fast Food Restaurant                 Field   
2   Empanada Restaurant            Donut Shop             Drugstore   
3   Fried Chicken Joint           Coffee Shop            Beer Store   
4    Athletics & Sports                   Pub                  Pool   

  8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
0            Restaurant                        Hotel    American Restaurant  
1        Farmers Market           Falafel Restaurant             Donut Shop  
2   Dumpling Restaurant  Eastern European Restaurant      Electronics Store  
3        Discount Store         Fast Food Restaurant            Pizza Place  
4           Coffee Shop                          Gym               Pharmacy

In [80]:
# cluster Neighbourhood
# set number of clusters
kclusters = 5

tor_grouped_clustering = tor_grouped_sum.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(tor_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 2, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

In [81]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood
# add clustering labels
tor_neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
tor_merged = df_toronto
tor_merged = tor_merged.join(tor_neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

tor_merged.head()

Neighbourhood   Latitude  Longitude  \
0                          Rouge, Malvern  43.806686 -79.194353   
1  Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497   
2       Guildwood, Morningside, West Hill  43.763573 -79.188711   
3                                  Woburn  43.770992 -79.216917   
4                               Cedarbrae  43.773136 -79.239476   

   Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0             1.0  Fast Food Restaurant           Yoga Studio   
1             0.0        History Museum                   Bar   
2             0.0           Pizza Place   Rental Car Location   
3             0.0           Coffee Shop     Korean Restaurant   
4             0.0                  Bank  Caribbean Restaurant   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
0           Event Space             Drugstore          Dumpling Restaurant   
1           Event Space   Dumpling Restaurant  Eastern European Restaurant   
2        Medical Center    Mexican Restaurant            Electronics Store   
3     Indian Restaurant            Donut Shop          Dumpling Restaurant   
4      Hakka Restaurant    Athletics & Sports                       Lounge   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Eastern European Restaurant     Electronics Store   Empanada Restaurant   
1            Electronics Store   Empanada Restaurant  Ethiopian Restaurant   
2                 Intersection        Breakfast Spot          Cupcake Shop   
3  Eastern European Restaurant     Electronics Store   Empanada Restaurant   
4              Thai Restaurant                Bakery   Fried Chicken Joint   

  9th Most Common Venue 10th Most Common Venue  
0  Ethiopian Restaurant     Falafel Restaurant  
1    Falafel Restaurant             Donut Shop  
2   Dumpling Restaurant                  Field  
3  Ethiopian Restaurant            Yoga Studio  
4   Empanada Restaurant    Dumpling Restaurant

In [82]:
tor_merged = tor_merged.fillna(0)

In [83]:
tor_merged['Cluster Labels'] = tor_merged['Cluster Labels'].apply(np.int32)

In [84]:
# visualize the resulting clusters
# create map
latitude = 43.653963
longitude = -79.387207
tor_map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(tor_merged['Latitude'], tor_merged['Longitude'], tor_merged['Neighbourhood'], tor_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(tor_map_clusters)
       
tor_map_clusters

### Get venues in Los Angeles

In [85]:
# Los Angeles venues
las_venues = getNearbyVenues(names=Los_df['Neighbourhood'],
                                   latitudes=Los_df['Latitude'],
                                   longitudes=Los_df['Longitude']
                                  )
print(las_venues.shape)

Adams_Normandie
Arleta
Arlington_Heights
Atwater_Village
Baldwin_Hills/Crenshaw
Bel-Air
Beverly_Crest
Beverly_Grove
Beverlywood
Boyle_Heights
Brentwood
Broadway_Manchester
Canoga_Park
Carthay
Central_Alameda
Century_City
Chatsworth
Chesterfield_Square
Cheviot_Hills
Chinatown
Cypress_Park
Del_Rey
Downtown
Eagle_Rock
East_Hollywood
Echo_Park
El_Sereno
Elysian_Park
Elysian_Valley
Encino
Exposition_Park
Fairfax
Florence
Glassell_Park
Gramercy_Park
Granada_Hills
Green_Meadows
Hancock_Park
Harbor_City
Harbor_Gateway
Harvard_Heights
Harvard_Park
Highland_Park
Historic_South_Central
Hollywood
Hollywood_Hills
Hollywood_Hills_West
Hyde_Park
Jefferson_Park
Koreatown
Lake_Balboa
Lake_View_Terrace
Larchmont
Leimert_Park
Lincoln_Heights
Los_Feliz
Manchester_Square
Mar_Vista
Mid_City
Mid_Wilshire
Mission_Hills
Montecito_Heights
Mount_Washington
North_Hills
North_Hollywood
Northridge
Pacific_Palisades
Pacoima
Palms
Panorama_City
Pico_Robertson
Pico_Union
Playa_del_Rey
Playa_Vista
Porter_Ranch
Rancho_P

In [86]:
las_venues.head(5)

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0  Adams_Normandie                34.03097               -118.30027   
1  Adams_Normandie                34.03097               -118.30027   
2  Adams_Normandie                34.03097               -118.30027   
3  Adams_Normandie                34.03097               -118.30027   
4  Adams_Normandie                34.03097               -118.30027   

               Venue  Venue Latitude  Venue Longitude    Venue Category  
0  Orange Door Sushi       34.032270      -118.299541  Sushi Restaurant  
1      Sushi Delight       34.032445      -118.299525  Sushi Restaurant  
2        Little Xian       34.032292      -118.299465  Sushi Restaurant  
3  Tacos La Estrella       34.032230      -118.300757        Taco Place  
4              Shell       34.033095      -118.300025       Gas Station

In [87]:
# analyze each Neighbourhood in Los Angeles
# one hot encoding
las_onehot = pd.get_dummies(las_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighbourhood column back to dataframe
las_onehot['Neighbourhood'] = las_venues['Neighbourhood'] 

# move neighbourhood column to the first column
fixed_columns = [las_onehot.columns[-1]] + list(las_onehot.columns[:-1])
las_onehot = las_onehot[fixed_columns]

las_onehot.head(5)

Neighbourhood  ATM  Accessories Store  Adult Boutique  \
0  Adams_Normandie    0                  0               0   
1  Adams_Normandie    0                  0               0   
2  Adams_Normandie    0                  0               0   
3  Adams_Normandie    0                  0               0   
4  Adams_Normandie    0                  0               0   

   American Restaurant  Arcade  Art Gallery  Art Museum  Arts & Crafts Store  \
0                    0       0            0           0                    0   
1                    0       0            0           0                    0   
2                    0       0            0           0                    0   
3                    0       0            0           0                    0   
4                    0       0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Australian Restaurant  Auto Workshop  \
0                 0                   0                      0              0   
1                 0                   0                      0              0   
2                 0                   0                      0              0   
3                 0                   0                      0              0   
4                 0                   0                      0              0   

   Automotive Shop  BBQ Joint  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0                0          0           0       0     0    0               0   
1                0          0           0       0     0    0               0   
2                0          0           0       0     0    0               0   
3                0          0           0       0     0    0               0   
4                0          0           0       0     0    0               0   

   Basketball Court  Beach  Bed & Breakfast  Beer Bar  Beer Garden  \
0                 0      0                0         0            0   
1                 0      0                0         0            0   
2                 0      0                0         0            0   
3                 0      0                0         0            0   
4                 0      0                0         0            0   

   Big Box Store  Bike Rental / Bike Share  Board Shop  Bookstore  Boutique  \
0              0                         0           0          0         0   
1              0                         0           0          0         0   
2              0                         0           0          0         0   
3              0                         0           0          0         0   
4              0                         0           0          0         0   

   Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
0               0        0            0                0         0   
1               0        0            0                0         0   
2               0        0            0                0         0   
3               0        0            0                0         0   
4               0        0            0                0         0   

   Burger Joint  Burrito Place  Bus Line  Bus Station  Bus Stop  \
0             0              0         0            0         0   
1             0              0         0            0         0   
2             0              0         0            0         0   
3             0              0         0            0         0   
4             0              0         0            0         0   

   Business Service  Butcher  Cafeteria  Café  Cajun / Creole Restaurant  \
0                 0        0          0     0                          0   
1                 0        0          0     0                          0   
2                 0        0          0     0                          0   
3                 0        0          0     0                          0   
4                 0        0          0     0                          0   

   Car

In [88]:
las_onehot.shape

(1943, 271)

There is 270 categories of venues in Los Angeles.

In [111]:
# count number of categories of venues in each neiborhood
las_onehot_group=las_onehot.groupby(['Neighbourhood']).sum()
las_onehot_group.head()

ATM  Accessories Store  Adult Boutique  \
Neighbourhood                                                    
Adams_Normandie           0                  0               0   
Arleta                    0                  0               0   
Arlington_Heights         0                  0               0   
Atwater_Village           0                  0               0   
Baldwin_Hills/Crenshaw    0                  0               0   

                        American Restaurant  Arcade  Art Gallery  Art Museum  \
Neighbourhood                                                                  
Adams_Normandie                           0       0            0           0   
Arleta                                    0       0            0           0   
Arlington_Heights                         0       0            0           0   
Atwater_Village                           1       0            0           0   
Baldwin_Hills/Crenshaw                    0       0            0           0   

                        Arts & Crafts Store  Asian Restaurant  \
Neighbourhood                                                   
Adams_Normandie                           0                 0   
Arleta                                    0                 0   
Arlington_Heights                         0                 0   
Atwater_Village                           0                 0   
Baldwin_Hills/Crenshaw                    0                 0   

                        Athletics & Sports  Australian Restaurant  \
Neighbourhood                                                       
Adams_Normandie                          0                      0   
Arleta                                   0                      0   
Arlington_Heights                        0                      0   
Atwater_Village                          1                      0   
Baldwin_Hills/Crenshaw                   0                      0   

                        Auto Workshop  Automotive Shop  BBQ Joint  Bagel Shop  \
Neighbourhood                                                                   
Adams_Normandie                     0                0          0           0   
Arleta                              0                0          0           0   
Arlington_Heights                   0                0          0           0   
Atwater_Village                     0                0          0           0   
Baldwin_Hills/Crenshaw              0                0          0           0   

                        Bakery  Bank  Bar  Baseball Field  Basketball Court  \
Neighbourhood                                                                 
Adams_Normandie              0     0    0               0                 0   
Arleta                       1     0    0               0                 0   
Arlington_Heights            0     0    0               0                 0   
Atwater_Village              1     1    2               0                 0   
Baldwin_Hills/Crenshaw       0     0    0               0                 0   

                        Beach  Bed & Breakfast  Beer Bar  Beer Garden  \
Neighbourhood                                                           
Adams_Normandie             0                0         0            0   
Arleta                      0                0         0            0   
Arlington_Heights           0                0         0            0   
Atwater_Village             0                0         0            0   
Baldwin_Hills/Crenshaw      0                0         0            0   

                        Big Box Store  Bike Rental / Bike Share  Board Shop  \
Neighbourhood                                                                 
Adams_Normandie                     0                         0           0   
Arleta                              0                         0           0   
Arlington_Heights                   0                         0           0   
Atwater_Village                

In [90]:
# Calculate the total per each category of venue in Los Angeles
las_city=pd.DataFrame(data=las_onehot_group.sum(axis=0))

In [91]:
las_city=las_city.transpose()

In [92]:
las_city.rename(index = {0:"Los Angeles"},inplace=True)

We would like to calculate number of venues under each type in Los Angeles.

In [93]:
# create a dataframe which contains the sum of venues by category in Los Angeles
las_city

ATM  Accessories Store  Adult Boutique  American Restaurant  \
Los Angeles    6                  2               1                   27   

             Arcade  Art Gallery  Art Museum  Arts & Crafts Store  \
Los Angeles       3            4           3                    9   

             Asian Restaurant  Athletics & Sports  Australian Restaurant  \
Los Angeles                17                   2                      1   

             Auto Workshop  Automotive Shop  BBQ Joint  Bagel Shop  Bakery  \
Los Angeles              1                4          4           5      36   

             Bank  Bar  Baseball Field  Basketball Court  Beach  \
Los Angeles    18   25               3                 2      4   

             Bed & Breakfast  Beer Bar  Beer Garden  Big Box Store  \
Los Angeles                2         2            1              3   

             Bike Rental / Bike Share  Board Shop  Bookstore  Boutique  \
Los Angeles                         1           1          5        10   

             Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  Building  \
Los Angeles              15        2            1                3         2   

             Burger Joint  Burrito Place  Bus Line  Bus Station  Bus Stop  \
Los Angeles            32              2         1            5         1   

             Business Service  Butcher  Cafeteria  Café  \
Los Angeles                 7        1          2    41   

             Cajun / Creole Restaurant  Carpet Store  Check Cashing Service  \
Los Angeles                          2             1                      2   

             Cheese Shop  Chinese Restaurant  Chocolate Shop  Clothing Store  \
Los Angeles            1                  34               1              25   

             Cocktail Bar  Coffee Shop  College Bookstore  College Classroom  \
Los Angeles             3           77                  1                  1   

             College Quad  College Residence Hall  College Theater  \
Los Angeles             1                       1                1   

             Comedy Club  Comic Shop  Community Center  Concert Hall  \
Los Angeles            1           1                 1             4   

             Construction & Landscaping  Convenience Store  Cosmetics Shop  \
Los Angeles                           4                 33              19   

             Costume Shop  Creperie  Cuban Restaurant  Cultural Center  \
Los Angeles             1         1                 3                1   

             Cycle Studio  Dance Studio  Deli / Bodega  Dentist's Office  \
Los Angeles             1             3              7                 1   

             Department Store  Design Studio  Dessert Shop  \
Los Angeles                 6              2             7   

             Dim Sum Restaurant  Diner  Discount Store  Dive Bar  Dog Run  \
Los Angeles                   2     11              10         7        1   

             Doner Restaurant  Donut Shop  Dry Cleaner  Dumpling Restaurant  \
Los Angeles                 1          23            1                    3   

             Electronics Store  Ethiopian Restaurant  Fabric Shop  \
Los Angeles                  2                     8            2   

             Falafel Restaurant  Farmers Market  Fast Food Restaurant  \
Los Angeles                   1               5                    37   

             Filipino Restaurant  Financial or Legal Service  Flower Shop  \
Los Angeles                    3                           1            7   

             Food  Food & Drink Shop  Food Court  Food Stand  Food Truck  \
Los Angeles    10                  2           2           1          34   

             Fountain  French Restaurant  Fried Chicken Joint  \
Los Angeles         1                  8                    7   

             Frozen Yogurt Shop  Furniture / Home Store  Garden  \
Los Angeles                   6                      12       3   

             Garden Center

In [94]:
#  group rows by neighborhood and by taking the mean of the frequency of occurrence of each category in Los Angeles
las_grouped_sum = las_onehot.groupby('Neighbourhood').mean().reset_index()
las_grouped_sum.head(5)

Neighbourhood  ATM  Accessories Store  Adult Boutique  \
0         Adams_Normandie  0.0                0.0             0.0   
1                  Arleta  0.0                0.0             0.0   
2       Arlington_Heights  0.0                0.0             0.0   
3         Atwater_Village  0.0                0.0             0.0   
4  Baldwin_Hills/Crenshaw  0.0                0.0             0.0   

   American Restaurant  Arcade  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0.000000     0.0          0.0         0.0                  0.0   
1             0.000000     0.0          0.0         0.0                  0.0   
2             0.000000     0.0          0.0         0.0                  0.0   
3             0.023256     0.0          0.0         0.0                  0.0   
4             0.000000     0.0          0.0         0.0                  0.0   

   Asian Restaurant  Athletics & Sports  Australian Restaurant  Auto Workshop  \
0               0.0            0.000000                    0.0            0.0   
1               0.0            0.000000                    0.0            0.0   
2               0.0            0.000000                    0.0            0.0   
3               0.0            0.023256                    0.0            0.0   
4               0.0            0.000000                    0.0            0.0   

   Automotive Shop  BBQ Joint  Bagel Shop    Bakery      Bank       Bar  \
0              0.0        0.0         0.0  0.000000  0.000000  0.000000   
1              0.0        0.0         0.0  0.166667  0.000000  0.000000   
2              0.0        0.0         0.0  0.000000  0.000000  0.000000   
3              0.0        0.0         0.0  0.023256  0.023256  0.046512   
4              0.0        0.0         0.0  0.000000  0.000000  0.000000   

   Baseball Field  Basketball Court  Beach  Bed & Breakfast  Beer Bar  \
0             0.0               0.0    0.0              0.0       0.0   
1             0.0               0.0    0.0              0.0       0.0   
2             0.0               0.0    0.0              0.0       0.0   
3             0.0               0.0    0.0              0.0       0.0   
4             0.0               0.0    0.0              0.0       0.0   

   Beer Garden  Big Box Store  Bike Rental / Bike Share  Board Shop  \
0          0.0            0.0                       0.0         0.0   
1          0.0            0.0                       0.0         0.0   
2          0.0            0.0                       0.0         0.0   
3          0.0            0.0                       0.0         0.0   
4          0.0            0.0                       0.0         0.0   

   Bookstore  Boutique  Breakfast Spot  Brewery  Bridal Shop  Bubble Tea Shop  \
0        0.0       0.0        0.000000      0.0          0.0              0.0   
1        0.0       0.0        0.000000      0.0          0.0              0.0   
2        0.0       0.0        0.000000      0.0          0.0              0.0   
3        0.0       0.0        0.023256      0.0          0.0              0.0   
4        0.0       0.0        0.000000      0.0          0.0              0.0   

   Building  Burger Joint  Burrito Place  Bus Line  Bus Station  Bus Stop  \
0       0.0           0.0            0.0       0.0          0.0       0.0   
1       0.0           0.0            0.0       0.0          0.0       0.0   
2       0.0           0.0            0.0       0.0          0.0       0.0   
3       0.0           0.0            0.0       0.0          0.0       0.0   
4       0.0           0.0            0.0       0.0          0.0       0.0   

   Business Service  Butcher  Cafeteria   Café  Cajun / Creole Restaurant  \
0          0.000000      0.0        0.0  0.000                        0.0   
1          0.000000      0.0        0.0  0.000                        0.0   
2          0.000000      0.0        0.0  0.125                        0.0   
3          0.023256      0.0        0.0  0.000        

In [95]:
# create a new dataframe
las_neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
las_neighborhoods_venues_sorted['Neighbourhood'] = las_grouped_sum['Neighbourhood']

for ind in np.arange(las_grouped_sum.shape[0]):
    las_neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(las_grouped_sum.iloc[ind, :], num_top_venues)

las_neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue 2nd Most Common Venue  \
0         Adams_Normandie      Sushi Restaurant                  Park   
1                  Arleta           Video Store         Historic Site   
2       Arlington_Heights            Donut Shop   Rental Car Location   
3         Atwater_Village            Food Truck                   Bar   
4  Baldwin_Hills/Crenshaw           Flower Shop           Pizza Place   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0         Grocery Store            Playground           Gas Station   
1                Bakery     Convenience Store           Flower Shop   
2         Grocery Store        Shop & Service                  Café   
3  Fast Food Restaurant    Chinese Restaurant     Mobile Phone Shop   
4        Clothing Store   Filipino Restaurant  Ethiopian Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0             Locksmith            Taco Place  Ethiopian Restaurant   
1    Seafood Restaurant  Ethiopian Restaurant           Fabric Shop   
2    Mexican Restaurant            Restaurant           Pizza Place   
3        Sandwich Place            Steakhouse        Breakfast Spot   
4           Fabric Shop    Falafel Restaurant        Farmers Market   

  9th Most Common Venue      10th Most Common Venue  
0           Fabric Shop          Falafel Restaurant  
1    Falafel Restaurant              Farmers Market  
2   Filipino Restaurant  Financial or Legal Service  
3                   Spa               Shopping Mall  
4  Fast Food Restaurant                 Yoga Studio

In [96]:
# cluster Neighbourhood
# set number of clusters
kclusters = 5

las_grouped_clustering = las_grouped_sum.drop('Neighbourhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(las_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 3, 4, 1, 1, 1], dtype=int32)

In [97]:
# create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood
# add clustering labels
las_neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)
las_merged = Los_df
las_merged = las_merged.join(las_neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

las_merged.head()

Neighbourhood   Longitude   Latitude  Cluster Labels  \
0         Adams_Normandie -118.300270  34.030970             1.0   
1                  Arleta -118.430015  34.240603             1.0   
2       Arlington_Heights -118.320109  34.043611             1.0   
3         Atwater_Village -118.265808  34.124908             1.0   
4  Baldwin_Hills/Crenshaw -118.366700  34.019090             1.0   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0      Sushi Restaurant                  Park         Grocery Store   
1           Video Store         Historic Site                Bakery   
2            Donut Shop   Rental Car Location         Grocery Store   
3            Food Truck                   Bar  Fast Food Restaurant   
4           Flower Shop           Pizza Place        Clothing Store   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0            Playground           Gas Station             Locksmith   
1     Convenience Store           Flower Shop    Seafood Restaurant   
2        Shop & Service                  Café    Mexican Restaurant   
3    Chinese Restaurant     Mobile Phone Shop        Sandwich Place   
4   Filipino Restaurant  Ethiopian Restaurant           Fabric Shop   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0            Taco Place  Ethiopian Restaurant           Fabric Shop   
1  Ethiopian Restaurant           Fabric Shop    Falafel Restaurant   
2            Restaurant           Pizza Place   Filipino Restaurant   
3            Steakhouse        Breakfast Spot                   Spa   
4    Falafel Restaurant        Farmers Market  Fast Food Restaurant   

       10th Most Common Venue  
0          Falafel Restaurant  
1              Farmers Market  
2  Financial or Legal Service  
3               Shopping Mall  
4                 Yoga Studio

In [98]:
las_merged = las_merged.fillna(0)

In [99]:
las_merged['Cluster Labels'] = las_merged['Cluster Labels'].apply(np.int32)

In [100]:
# visualize the resulting clusters
# create map
latitude = 34.052235
longitude = -118.243683
las_map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(las_merged['Latitude'], las_merged['Longitude'], las_merged['Neighbourhood'], las_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(las_map_clusters)
       
las_map_clusters

### Combine three cities summary data together

We are going to compare three cities using cosine similarity. First I will create a dataframe that combine all three cities venues and categories together.

In [101]:
all_city=pd.concat([ny_city,tor_city,las_city], axis=0, ignore_index=True, sort=False)
all_city.rename(index = {0:"New York",
                         1:"Toronto",
                         2:"Los Angeles"},inplace=True)

In [102]:
all_city

Accessories Store  Adult Boutique  Afghan Restaurant  \
New York                    12             3.0                3.0   
Toronto                      1             NaN                1.0   
Los Angeles                  2             1.0                NaN   

             African Restaurant  Airport Terminal  American Restaurant  \
New York                    6.0               1.0                  169   
Toronto                     NaN               2.0                   29   
Los Angeles                 NaN               NaN                   27   

             Animal Shelter  Antique Shop  Arcade  Arepa Restaurant  \
New York                1.0          13.0     3.0              10.0   
Toronto                 NaN           2.0     NaN               NaN   
Los Angeles             NaN           NaN     3.0               NaN   

             Argentinian Restaurant  Art Gallery  Art Museum  \
New York                        9.0           55           7   
Toronto                         NaN           13           2   
Los Angeles                     NaN            4           3   

             Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
New York                      15                   2.0                51   
Toronto                        7                   NaN                17   
Los Angeles                    9                   NaN                17   

             Athletics & Sports  Auditorium  Australian Restaurant  \
New York                     16         1.0                    4.0   
Toronto                       6         NaN                    NaN   
Los Angeles                   2         NaN                    1.0   

             Austrian Restaurant  Auto Garage  Auto Workshop  Automotive Shop  \
New York                     2.0          1.0              3              3.0   
Toronto                      NaN          NaN              1              NaN   
Los Angeles                  NaN          NaN              1              4.0   

             BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
New York            27         1.0         106     231   154  210   
Toronto              5         1.0           7      49    18   42   
Los Angeles          4         NaN           5      36    18   25   

             Baseball Field  Baseball Stadium  Basketball Court  Bath House  \
New York                 20               3.0              13.0         1.0   
Toronto                   5               2.0               NaN         NaN   
Los Angeles               3               NaN               2.0         NaN   

             Beach  Beach Bar  Bed & Breakfast  Beer Bar  Beer Garden  \
New York        45        1.0                2        13         14.0   
Toronto          1        NaN                1        18          NaN   
Los Angeles      4        NaN                2         2          1.0   

             Beer Store  Big Box Store  Bike Rental / Bike Share  Bike Shop  \
New York            6.0            5.0                       6.0        4.0   
Toronto             9.0            NaN                       NaN        2.0   
Los Angeles         NaN            3.0                       1.0        NaN   

             Bike Trail  Bistro  Board Shop  Boat or Ferry  Bookstore  \
New York            3.0     8.0        10.0           10.0         52   
Toronto             NaN     4.0         NaN            1.0         18   
Los Angeles         NaN     NaN         1.0            NaN          5   

             Boutique  Bowling Alley  Boxing Gym  Brazilian Restaurant  \
New York           38           10.0        10.0                   5.0   
Toronto             5            NaN         NaN                   3.0   
Los Angeles        10            NaN         NaN                   NaN   

             Breakfast Spot  Brewery  Bridal Shop  Bridge  Bubble Tea Shop  \
New York                 38       14            3     3.0               30   
Toronto                  25       14

In [104]:
# replace NaN with 0
all_city = all_city.fillna(0)

Create a dataframe that contians three cities' vector.

In [105]:
all_city.shape

(3, 476)

In [107]:
all_city.astype(int)

Accessories Store  Adult Boutique  Afghan Restaurant  \
New York                    12               3                  3   
Toronto                      1               0                  1   
Los Angeles                  2               1                  0   

             African Restaurant  Airport Terminal  American Restaurant  \
New York                      6                 1                  169   
Toronto                       0                 2                   29   
Los Angeles                   0                 0                   27   

             Animal Shelter  Antique Shop  Arcade  Arepa Restaurant  \
New York                  1            13       3                10   
Toronto                   0             2       0                 0   
Los Angeles               0             0       3                 0   

             Argentinian Restaurant  Art Gallery  Art Museum  \
New York                          9           55           7   
Toronto                           0           13           2   
Los Angeles                       0            4           3   

             Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
New York                      15                     2                51   
Toronto                        7                     0                17   
Los Angeles                    9                     0                17   

             Athletics & Sports  Auditorium  Australian Restaurant  \
New York                     16           1                      4   
Toronto                       6           0                      0   
Los Angeles                   2           0                      1   

             Austrian Restaurant  Auto Garage  Auto Workshop  Automotive Shop  \
New York                       2            1              3                3   
Toronto                        0            0              1                0   
Los Angeles                    0            0              1                4   

             BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
New York            27           1         106     231   154  210   
Toronto              5           1           7      49    18   42   
Los Angeles          4           0           5      36    18   25   

             Baseball Field  Baseball Stadium  Basketball Court  Bath House  \
New York                 20                 3                13           1   
Toronto                   5                 2                 0           0   
Los Angeles               3                 0                 2           0   

             Beach  Beach Bar  Bed & Breakfast  Beer Bar  Beer Garden  \
New York        45          1                2        13           14   
Toronto          1          0                1        18            0   
Los Angeles      4          0                2         2            1   

             Beer Store  Big Box Store  Bike Rental / Bike Share  Bike Shop  \
New York              6              5                         6          4   
Toronto               9              0                         0          2   
Los Angeles           0              3                         1          0   

             Bike Trail  Bistro  Board Shop  Boat or Ferry  Bookstore  \
New York              3       8          10             10         52   
Toronto               0       4           0              1         18   
Los Angeles           0       0           1              0          5   

             Boutique  Bowling Alley  Boxing Gym  Brazilian Restaurant  \
New York           38             10          10                     5   
Toronto             5              0           0                     3   
Los Angeles        10              0           0                     0   

             Breakfast Spot  Brewery  Bridal Shop  Bridge  Bubble Tea Shop  \
New York                 38       14            3       3               30   
Toronto                  25       14

In [108]:
#Compute Cosine Similarity
from sklearn.metrics.pairwise import cosine_similarity
print(cosine_similarity(all_city, all_city))

[[1.         0.75198537 0.85064902]
 [0.75198537 1.         0.77485581]
 [0.85064902 0.77485581 1.        ]]


## Results

compare the cosine similarity bewteen each two cities. The cosine similarity value is between [0,1].

From the result above we see that the cosine similarity vaule between New York City and Toronto is 0.751985. 

Cosine similarity vaule between New York City and Los Angeles is 0.85065. 

Cosine similarity vaule between Toronto and Los Angeles is 0.77486. 

By look at the clusters of neibourhoods in these three cities, that New York City and Los Angeles also show similar cluster pattern then comparing them to Toronto.

## Discussion

Our analysis shows that New York City has much more categories of venues than Toronto and Los Angeles. In addition, New York City has much more restaurants than Toronto and Los Angeles. In general, New York City has larger amount of venues than Toronto and Los Angeles. Comapre the amount of venues and neibourhoods that Toronto and Los Angeles are more similar to each other. On the other hand, after we create cluster of neibourhood, it shows New York City and Los Angeles are more similar when compare the density of venues than compare to Toronto.After applying cosine similarity on three city vector, the calculation shows that New York City is more similar to Los Angeles than Toronto. In addition, to compare similarity between cities can also consider other factor more than venues, for example: housing price, population, income, density, etc. Also, there are more techniques to compare similarity between cities.

## Conclusion

Purpose of this project was to identify how similar are these three cities by comparing venue categories and density.We have presented the results of city similarities as well as an analysis on the frequencies of different venue categories and how each representation may affect them. By calculating density distribution from Foursquare data we have first identified general boroughs that justify further analysis. Clustering of those locations was then performed in order to see the distribution of different types of venues in each city. From the comparison result, we can conclude that New York City and Los Angeles are more similar than compare to Toronto.